synapse/api/room_versions.py

```diff
@@ -208,5 +208,39 @@ class RoomVersions:
         RoomVersions.MSC3083,
         RoomVersions.V7,
     )
-    # Note that we do not include MSC2043 here unless it is enabled in the config.
 }  # type: Dict[str, RoomVersion]
+
+
+@attr.s(slots=True, frozen=True)
+class RoomVersionCapability:
+    """An object which describes the unique attributes of a room version."""
+
+    identifier = attr.ib(type=str)  # the identifier for this capability
+    preferred_version = attr.ib(type=Optional[RoomVersion])
+    support_check_lambda = attr.ib(type=Callable[[RoomVersion], bool])
```

Comment:
We're starting to use native type hints in new code:

```suggestion
@attr.s(slots=True, frozen=True, auto_attribs=True)
class RoomVersionCapability:
    """An object which describes the unique attributes of a room version."""

    identifier: str  # the identifier for this capability
    preferred_version: Optional[RoomVersion]
    support_check_lambda: Callable[[RoomVersion], bool])
```